File: 0126-word-ladder-ii/0126-word-ladder-ii.cpp
class Solution {
public:
    void dfs(unordered_map<string,int>& dist,string targetWord, string startWord,vector<string> res, vector<vector<string>>& ans) {
        

        if(targetWord==startWord) {
            reverse(res.begin(),res.end());
            ans.push_back(res);
            reverse(res.begin(),res.end());
            return;
        }

        string v = startWord;
        int sz = v.size();
        for(int i = 0;i<sz;i++) {
            for(char ch = 'a';ch<='z';ch++) {
                v[i] = ch;
                if(dist.count(v) && dist[v]+1 == dist[startWord]) {
                    res.push_back(v);
                    dfs(dist,targetWord,v,res,ans);

                    res.pop_back();
                }
            }
            v = startWord;
        }

    }
    vector<vector<string>> findLadders(string startWord, string targetWord, vector<string>& wordList) {
        unordered_set<string> dict(wordList.begin(),wordList.end());
        unordered_map<string,int> dist;
        vector<vector<string>> ans;
        
        dist[startWord] = 1;


        
       
        if(dict.count(targetWord)==0) return ans;
       
        
        queue<string> que;
        que.push(startWord);
        dict.erase(startWord);
        while(!que.empty()) {
            string u = que.front();
            
            que.pop();
            if(targetWord == u) break;
            int len = u.length();
            for(int i = 0;i<len;i++) {
                string temp = u;
                for(char ch = 'a';ch<='z';ch++) {
                    if(ch==u[i]) continue;
                    temp[i] = ch;
                    if(dict.count(temp)!=0) {
                    
                            dist[temp] = dist[u]+1;
                            dict.erase(temp);
                            que.push(temp);
                         
                        
                    }
                }
            }
            
        }
        vector<string> res;
        res.push_back(targetWord);
        dfs(dist,startWord,targetWord,res,ans);
        return ans;
    }
};

File: 0126-word-ladder-ii/0126-word-ladder-ii.java
class Solution {

    public void dfs(String beginWord, String endWord, 
                    Map<String,Integer> dist, List<List<String>> ans,
                    List<String> anslist) {

        if(beginWord.equals(endWord)) {
            ArrayList<String> tmp = new ArrayList<>(anslist);
            Collections.reverse(tmp);
            ans.add(tmp);
            return;
        }

        int level = dist.get(beginWord);

        int n = beginWord.length();

        for(int i = 0;i<n;i++) {
            StringBuilder sb = new StringBuilder(beginWord);
            
            for(char ch='a';ch<='z';ch++) {
                if(ch==beginWord.charAt(i)) continue;
                sb.setCharAt(i,ch);
                String modString = sb.toString();
                int levelDist = dist.getOrDefault(modString,-1);
                if(level-levelDist==1) {
                    anslist.add(modString);
                    dfs(modString,endWord,dist,ans,anslist);
                    anslist.remove(modString);
                }
            }

            sb.setCharAt(i,beginWord.charAt(i));
        }
        

    }

    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {

        List<List<String>> ans = new ArrayList<>();
        Set<String> wordSet = new HashSet<>();
        Map<String,Integer> dist = new HashMap<>();
        for(String s: wordList) wordSet.add(s);
        if(wordSet.contains(endWord)==false) return ans;
        int level = ladderLength(beginWord,endWord,wordList,dist);
        List<String> anslist = new ArrayList<>();
        anslist.add(endWord);

        dfs(endWord,beginWord,dist,ans,anslist);

        return ans;
        
    }

     public int ladderLength(String beginWord, String endWord, List<String> wordList, Map<String,Integer> dist) {
       
        // Map<String,Integer> dist = new HashMap<>();

        for(String s: wordList) dist.put(s,Integer.MAX_VALUE);
        
        if(dist.containsKey(endWord)==false) return 0;

        Queue<String> que = new ArrayDeque<>();
        que.add(beginWord);
        dist.put(beginWord,1);

        while(!que.isEmpty()) {
            StringBuilder u = new StringBuilder(que.poll());
            String ustr = u.toString();

            int sn = u.length();
            for(int i = 0;i<sn;i++) {
                
                char reserv = u.charAt(i);

                for(char ch='a';ch<='z';ch++) {
                    if(ch==reserv) continue;
                    u.setCharAt(i,ch);
                    String v = u.toString();
                    if(dist.containsKey(v)==true && dist.get(v)==Integer.MAX_VALUE) {
                        dist.put(v,dist.get(ustr)+1);
                        if(v.equals(endWord)==true) {
                            return dist.get(v);
                        }
                        que.add(v);
                    }
                }

                u.setCharAt(i,reserv);
            }
        }

        return 0;
    }
}